### ISSUE TITLE
Real time data

### ISSUE THREAD

**Issue by dhruv-jain** [metasys-server/api-landing#20]
Is it possible to get real time utilities data from metasys api? Thank you.

**Comment by mgwelch** [comment]
At 10.0 it's not possible to get live data. The closest approximation is to create a Trend extension on the appropriate objects. Set the sample time to some period that is small enough (say 5 minutes) and then request just a page size of 1 to get the latest sample.

We are working at exposing the ability to read live values from the system in a future release.

**Comment by dhruv-jain** [comment]
Thank you, I will try that.

**Comment by mgwelch** [comment]
Excellent. Stay tuned for future releases.

**Comment by dhruv-jain** [comment]
What command can you query to get trend data on time intervals. Such as the 5 latest minutes on a specific point?

`GET /api/objects/b887ceaa-d1c7-5818-b689-1435161bb323/attributes/85/samples?startTime=2018-08-27&endTime=2018-08-29&page=1&pageSize=1000&sort=-timestamp HTTP/1.1`   - How do get past 5 minutes instead of using days?

**Comment by mgwelch** [comment]
You can specify a full date time is ISO8601 format. For example: 2019-07-25T20:14:37Z

**Comment by Austin-TheTrueShinobi** [comment]
Out of curiosity. Is there a script or other documentation besides what is on the metasys.github.io? I'm still learning the architecture of the API and the process of commands to call to get the most accurate trend data on each endpoint. (That closely resembles live data)

I found this. 
![image](https://user-images.githubusercontent.com/50872549/61964003-e3039d00-af9a-11e9-8b50-6d66dceb2e88.png)

Can you enlighten me on how trend data is created and what role "objects" play in capturing endpoint data?

**Comment by mgwelch** [comment]
>  Is there a script or other documentation besides what is on the metasys.github.io for capturing endpoint data?

@Austin-TheTrueShinobi we've identified the need for a "Metasys Domain Model" guide which would cover objects, points, spaces, equipment and all of their relationships.

We also plan to have additional tutorials and guides for handling objects. The documentation you sited is actually for Metasys 10.1 (not Metasys 10.0). 

> Is there a script or other documentation besides what is on the metasys.github.io for capturing endpoint data?
> Can you enlighten me on how trend data is created and what role "objects" play in capturing endpoint data?

There are two different concepts we are talking about in this issue

1. Trend samples (for a given attribute of a given object)
2. Current value (for a given attribute of a given object)

In Metasys, samples (trend data) are only taken on objects that have "Trend Extensions" defined for them (Newer versions of Metasys allow for setting up trending without the need of a Trend Extension but still require that sampling be configured). When trending is setup on an object you define what attribute of the object you want to trend (typically the "presentValue" attribute) and how often you want to sample. If you have a server on your site you can configure trends to forward to the database on the server for long term storage.

The endpoint you highlighted above is appropriate for reading the current value of the `presentValue` attribute of the object with the given `id`. It doesn't actually look in the database at all for this value, but actually forwards a request to the appropriate network engine to request the value.

**Comment by Austin-TheTrueShinobi** [comment]
Thus, retrieved attribute data from the "samples section" i.e.
![image](https://user-images.githubusercontent.com/50872549/62051839-39165180-b1e2-11e9-9340-265391ce4710.png)

requires a trend extension to be in place. - in which you can find possible values in /points/{id}/trendedAttributes. But to request trend data from the objects directly, such as in our example, sampling still must be configured for the specified pull? -Is that just an additional trending pull in the Metasys application? 

-**To summarize**: whats the difference between pulling from the "samples" vs. pulling trends from the objects like in the example; /api/objects/b887ceaa-d1c7-5818-b689-1435161bb323/attributes/85/samples?startTime=2018-08-27&endTime=2018-08-29&page=1&pageSize=1000&sort=-timestamp 

Currently our trends are being forwarded into our DB on site. However, I am just trying to extract the "most live" object attribute data directly so i can parse it into a local JSON/.txt file. 

> The endpoint you highlighted above is appropriate for reading the current value of the presentValue attribute of the object with the given id. It doesn't actually look in the database at all for this value, but actually forwards a request to the appropriate network engine to request the value.
So is this the "live" value of that object? 

Thanks for the help!!

**Comment by mgwelch** [comment]
@Austin-TheTrueShinobi I'm not sure I understand your question. Let me step back and describe more of what's going on.

1. Trending can be configured for an attribute of an object.
2. When trends are configured, samples start to be buffered locally on the engine for the specified attribute(s). This results in samples that are on the engine but not in the database on the server.
3. Eventually samples from engine buffers are forwarded to the server. This results in samples being in the database on the server, and the buffers on the engines being emptied to make room for more local samples.
4. Separate from trending/sampling you can read the current value of an attribute.

Our API at 10.0 allows you to read trend samples from the database on the server. It doesn't allow you to read samples that are still on the engine.

At 10.1 you can read the live value of an attribute on an engine.

There are various ways to configure trends to forward to the server so that the server gets those samples more quickly. I believe your team is in communication with our field support team. They should be able to help address any question about forwarding you might have to see if they meet your needs.

Did this answer your question?

**Comment by Austin-TheTrueShinobi** [comment]
That clarified most of my current questions extremely well! Thank you.

The last question I have for now is in regards to getting the object id's. Can you give me an example of how to URL encode the request to find the object id's?

**Such As:** "ClemsonADX:DILLARD-NCE/FCB.CHW BTU METER.AV-1" being the fully qualified name.

I try running: "curl -L -H "Authorization: Bearer ..." https://example.com/api/v2/objectIdentifiers?fqr=ClemsonADX%3ADILLARD-NCE%2FFCB.CHW%20BTU%20METER.AV-1 "
but to no avail. It does not return any JSON response. 

Thanks again!

**Comment by mgwelch** [comment]
@Austin-TheTrueShinobi the /objectIdentifiers endpoint wasn't available at 10.0 unfortunately.

I've written a little nodejs app at 10.0 that you could use to look up individual object identifiers by using the /objects endpoint: https://github.com/metasys-server/objectjs

If you have lots of points to lookup then you are best off just fetching all objects and storing them offline for future reference. Once you have the objects you can search your offline data for a given FQR and in the object payload you'll see the `id`.

**Comment by Austin-TheTrueShinobi** [comment]
I believe we are running 10.1 so I thought It might have to do with some syntactical error on the URL encoding part? I see we are running 10.1 in the UI

**Comment by mgwelch** [comment]
@Austin-TheTrueShinobi OK, if you have 10.1 then let me take a closer look at your example and I'll see if I can find the issue.

**Comment by Austin-TheTrueShinobi** [comment]
**Update:** I am now able to obtain an Object ID using the "/objectIdentifiers?" command that returns a response in the form of a lengthy string. But when I try to use the ID to gather information on a AV it returns with no response. 

I am trying to call the presentValue attribute on the object using
![image](https://user-images.githubusercontent.com/50872549/62137263-a6de7e00-b2b3-11e9-8303-87f24d78908c.png)

I am able to use the objectID to call the **https://hostname/api/v2/objects/id/points?** command and retrieve the attributeUrl of the object but it only give the description of "Present Value" and not the value.

Am i doing something incorrectly in my command? Or perhaps I need elevated privileges? 

**Edit:** both _https://example.com/api/v2/objects/id?includeSchema=false_ and _https://example.com/api/v2/objects/id/presentValue_ return with no response.

**Comment by mgwelch** [comment]
> Edit: both https://example.com/api/v2/objects/id?includeSchema=false and

What status code is being returned? You need a license to use the part of the API that lets you read live attribute values.

> https://example.com/api/v2/objects/id/presentValue

The API for getting an attribute would look like this (you are missing the attributes part of the path) `https://hostname/api/v2/objects/{id}/attributes/presentValue`

**Comment by Austin-TheTrueShinobi** [comment]
Hmm, I tried that command and It also returned with no status code. It just executes and returns with nothing. I'll check both privileges and the license and get back to you. -Thanks!

**Comment by mgwelch** [comment]
> Hmm, I tried that command and It also returned with no status code.

Use the `-i` switch with curl to get the response headers. Then you can see the status code.

> Edit: If we are running 10.1 do we not already have that licence?

No, not necessarily. It's a separately licensable component.

**Comment by Austin-TheTrueShinobi** [comment]
Here is the response header for a working command:
![points](https://user-images.githubusercontent.com/50872549/62220766-9b0bbe00-b37e-11e9-838b-ea26fd85587f.PNG)
**Running:** _https://hostname/api/v2/objects/objectID/points?_

Here are the response headers for the non-working commands:
![objects](https://user-images.githubusercontent.com/50872549/62220911-d27a6a80-b37e-11e9-941c-a754db5360b3.PNG)
**Running:** _https://hostname/api/v2/objects/objectID?_

![objectAttribute](https://user-images.githubusercontent.com/50872549/62221014-f938a100-b37e-11e9-80b1-f55281dd4092.PNG)
**Running:** _https://hostname/api/v2/objects/objectID/attributes/presentValue_

**Comment by mgwelch** [comment]
Yeah, that seems to confirm you are not licensed. I'm double checking that the error code 403 is returned when not licensed. I refer you to your local Johnson Controls branch for information on getting a license.

**Comment by Austin-TheTrueShinobi** [comment]
My go-to local Johnson Controls guy is on vacation this week so I suppose ill have to wait for him to get back. However If a license is what we need then that is what we will get if you would recommend it. 

Also, In regards to the trend-samples we discussed earlier. When the engine forwards the "samples" (collection of present values or any other attribute) to the DB server, does is just send a "trended value" of the samples to the server? Or does it send the whole collection of samples? 

**Additionally**, is there any way to call a trended sample for a time range that is not configured? i.e. Can I call a trend off of the engine without it being recorded in the DB? 

Thanks for your patience!

**Comment by mgwelch** [comment]
> My go-to local Johnson Controls guy is on vacation this week so I suppose ill have to wait for him to get back. However If a license is what we need then that is what we will get if you would recommend it.

If you want to use that API you'll need a license. You can ask for a trial license so you can explore the API first.

> Additionally, is there any way to call a trended sample for a time range that is not configured? i.e. Can I call a trend off of the engine without it being recorded in the DB?

The samples endpoint will only retrieve data from the DB. (The current release was really meant for accessing historical data and therefore wasn't concerned with retrieving the most recent data that is still on the engine as that can be configured to forward once a day or perhaps even more often). Also, there won't be any samples (on the engine or the DB) unless trending had been configured.

**Comment by Austin-TheTrueShinobi** [comment]
So when it forwards values to the DB, does it send ONE trended value that represents the collection of samples for your trend interval or does it just send the whole collection after whatever interval of time you set the system to pull? 

For example; If you set up your trend interval on an object to pull every 15 minutes. Every 15 minutes ONE trended value of the samples collection will be sent to the DB and recorded?

**Comment by mgwelch** [comment]
> So when it forwards values to the DB, does it send ONE trended value that represents the collection of samples for your trend interval or does it just send the whole collection after whatever interval of time you set the system to pull?

Good question. There are several things that can control how much information is sent when samples are forwarded to the server. Best to discuss this with your JCI contact when he's back to get the best advice for your situation. 

In general every trended attribute is going to collect a batch of samples. Then when the time comes to forward, all samples for all trended attributes will be sent at once. This is all impacted by the buffer size used for a trend configuration and the forwarding scheme used. It could be setup to send when buffer is full, or periodically at a certain time of the day. I'm not completely up to date on the trend configuration options. Again, your branch and our field support org would be best to answer these questions for you.

**Comment by Austin-TheTrueShinobi** [comment]
Hello again, I have talked to my support tech and have a few questions. 

**Currently** we are running the alpha version of 10.1 in which we assumed came with all the licenses offered. Is it possible that there is a bug in the software causing the commands to not return data? Or are we sure about the response header error codes being a licensing issue? 

We are in the process of moving to the beta version of 10.1 over the next few weeks so I wanted to be sure about what questions to ask in order to get what we need. 

Thanks as always

**Comment by mgwelch** [comment]
@Austin-TheTrueShinobi I'm reaching out to those that manage the Alpha program to find out the state of the licenses issued at your site. You can also use the licensing app on your server to check what licenses are installed.

> Is it possible that there is a bug in the software causing the commands to not return data? 

It's a possibility, but not something we've seen in our testing.

> Or are we sure about the response header error codes being a licensing issue?

The other possibility is that the user you are using to login doesn't have permissions for the data you are trying to read. Check the category of the objects you are trying to read against the permissions the user has.

In any case, I've got an email out to the alpha management to see if they can  tell us if they think a license was installed.

**Comment by Austin-TheTrueShinobi** [comment]
> The other possibility is that the user you are using to login doesn't have permissions for the data you are trying to read. Check the category of the objects you are trying to read against the permissions the user has.

Just ran with elevated privileges to no avail. :(

> In any case, I've got an email out to the alpha management to see if they can tell us if they think a license was installed.

Great, thanks! Let me know what you find out. In the meantime I will look into the licensing app on the server.

**Comment by Austin-TheTrueShinobi** [comment]
**Update:** Just checked the License app on the server... There is an option called "monitoring data and commanding the API" that is not licensed. Suppose that's the problem? :) 🥇

**Comment by mgwelch** [comment]
@Austin-TheTrueShinobi I expect someone to contact your university with licensing information. I'm on vacation this week, but I checked my email to see they are placing the order for you for the license.

**Comment by Austin-TheTrueShinobi** [comment]
Thanks for all your help. Yes, we got the license setup and are currently pulling data. Out of curiosity, is the API designed to handle multiple execution calls? Meaning, is there a way to make sequential or non-sequential calls to the API to get responses faster for larger call sets? 

### For example, if i were to call a presentValue on a set of 10,000 objectID's at
![time](https://user-images.githubusercontent.com/50872549/62781462-114aa780-ba86-11e9-96f8-fd99fa324edb.PNG)

Sequentially this would take about 25 minutes. Am I missing something? What is the different between the API retrieval process and the BACnet Metasys UI to SQL process in regards to speed? 

p.s. We can bring this conversation to a new issue thread if you would like.

**Comment by mgwelch** [comment]
@Austin-TheTrueShinobi 

You can however make calls in parallel. You'll want to measure the load on the server to be sure that you aren't impacting your other users.

You may find that initial reads of some attributes are quite a bit slower than what you posted. But there are some enhancements running in the background for items that have been read recently.

**Comment by Austin-TheTrueShinobi** [comment]
Do you know what that limit is? Also, out of curiosity do you happen to have a script that gets all objectID's for a given network engine? 

Thanks again! and sorry for the late reply.

**Comment by mgwelch** [comment]
@Austin-TheTrueShinobi our field support office should be able to lookup the limits for you. I believe we use standard IIS Modules so you can adjust to to meet your particular needs using IIS tooling.

To get all of the objects (and thereby their object ids) you'll want to call the /networkDevices API (/networkDevices/{id}/objects)

I don't have a script that does that but I have a script that gets one object id. It could be easily modified to do what you said. (But it's javascript): https://github.com/metasys-server/objectjs

**Comment by Austin-TheTrueShinobi** [comment]
Hello again! We are undergoing server upgrades and OS migrations so I've been waiting a bit to see what type of structure I wanted to build so apologies for the late reply! In regards to calling the API in parallel, what type/structure of parallel calls would you recommend? Things like GNU parallel and others are limited to CPU cores. I know of <Tasks>, <asynch>, and <Threads> but I'm still pretty new to Parallelism in computational logic.

Thanks as always

**Comment by mgwelch** [comment]
@Austin-TheTrueShinobi  What language are you using? If you are using .Net then research async/await and Task<T>. Most I/O methods in the framework are async by default. That means that your thread is not blocked while the call is being made. This means you can go ahead and make 1 or more calls. The framework then has different ways for you to wait for multiple requests to complete. For example this article shows a pretty simple way of awaiting on 3 outstanding requests. https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/how-to-make-multiple-web-requests-in-parallel-by-using-async-and-await

But there are a lot of ways for you to write this code.

**Comment by Austin-TheTrueShinobi** [comment]
Thanks! If i were to be running this code as a service/daemon on linux is there a language you would recommend for that?

**Comment by mgwelch** [comment]
The two frameworks that I'm most familiar with that you could use on linux are [dotnet core](https://dotnet.microsoft.com/download/dotnet-core) and [nodejs](https://nodejs.org/en/). Nodejs also has async/await syntax and has Promises (which are similar to the Task class in dotnet).

So that would be C# or JavaScript.

What language have you been using/testing with? Or have you just been using something like Postman?

**Comment by Austin-TheTrueShinobi** [comment]
Up until now I've been trying to use c# due to the fact that I'm familiar with creating service and the like in Visual Studio. I have also used a Bash command called "Parallel" that you can install to run complex curl commands. But I'm proficient in C/C++/Python/Java/bash/http and some other languages so I'm trying to find the more flexible solution that would run on Linux. 

Thanks though! I'll check those out.

**Comment by mgwelch** [comment]
I hope things are going well. I'm going to close this issue. Feel free to open a new one if needed. I think we've sufficiently discussed "real time data" for now.
